### PR TITLE
Reduce binary size

### PR DESCRIPTION
**Pull request by etiennebacher**

### CODE REVIEW

File: src/rust/src/lib.rs [hunk 1]
```diff
@@ -73,7 +76,7 @@ impl UnicodePosition {
 
 #[derive(Clone)]
 pub struct SgRoot {
-    inner: AstGrep<StrDoc<SupportLang>>,
+    inner: AstGrep<StrDoc<tree_sitter_facade_sg::Language>>,
```
Comment (HerringtonDarkholme) [inline src/rust/src/lib.rs:79]: IMHO it should be `StrDoc<R>` here. `StrDoc` accepts a type parameter satisfying `ast_grep_core::Language`.  The `R` language implemented here should be the one to use.

File: src/rust/src/lib.rs [hunk 2]
```diff
@@ -82,9 +85,9 @@ pub struct SgRoot {
 impl SgRoot {
     fn new(src: &str) -> Self {
         let position = UnicodePosition::new(src);
-        let lang = SupportLang::R;
```
Comment (HerringtonDarkholme) [inline src/rust/src/lib.rs:85]: ditto, `R.ast_grep(src)` should be sufficient here.

File: src/rust/src/language.rs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    fn get_ts_language(&self) -> TSLanguage;
+    /// ignore trivial tokens in language matching
+    fn skippable_kind_ids(&self) -> &'static [u16] {
+        &[]
+    }
+
+    /// normalize pattern code before matching
+    /// e.g. remove expression_statement, or prefer parsing {} to object over block
+    fn pre_process_pattern<'q>(&self, query: &'q str) -> Cow<'q, str> {
+        Cow::Borrowed(query)
```
Comment (HerringtonDarkholme) [inline src/rust/src/language.rs:48]: You may also need to pre_process_pattern if `$` is not a valid identifier in R.

File: src/rust/src/language.rs [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    // trailing anonymous multiple
+    let sigil = if dollar_count == 3 { expando } else { '$' };
+    ret.extend(repeat(sigil).take(dollar_count));
+    std::borrow::Cow::Owned(ret.into_iter().collect())
+}
+/// Trait to abstract ts-language usage in ast-grep, which includes:
+/// * which character is used for meta variable.
+/// * if we need to use other char in meta var for parser at runtime
+/// * pre process the Pattern code.
+pub trait Language: Clone {
```
Comment (HerringtonDarkholme) [inline src/rust/src/language.rs:55]: Why not `use ast_grep_core::Language` here?

**Comment by etiennebacher** [comment]
I think my last commit should work, thanks so much for your help @HerringtonDarkholme, I really appreciate!

**Review by HerringtonDarkholme — COMMENTED** [review]
Looks Nice!

**Pull request merged**